geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroSchemas.scala

```diff
@@ -0,0 +1,253 @@
+package org.locationtech.geomesa.features.avro.complex
+
+import org.apache.avro.Schema.Field
+import org.apache.avro.Schema.Type._
+import org.apache.avro.io.{Decoder, Encoder}
+import org.apache.avro.{Schema, SchemaBuilder}
+import org.geotools.feature.NameImpl
+import org.geotools.gml3.GMLSchema
+import org.geotools.xs.XSSchema
+import org.locationtech.geomesa.utils.geotools.RichAttributeDescriptors._
+import org.opengis.feature.`type`._
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+import scala.collection.immutable.{BitSet, Queue}
+import scala.collection.mutable
+
+object AvroSchemas {
+  val FEATURE_ID_AVRO_FIELD_NAME = "__fid__"
+  val AVRO_SIMPLE_FEATURE_VERSION = "__version__"
+  val AVRO_SIMPLE_FEATURE_USERDATA = "__userdata__"
+  val VERSION = 4
+  val AVRO_NAMESPACE = "org.geomesa"
+  val NAME_SIMPLE_CONTENT = new NameImpl(AVRO_NAMESPACE, "simpleContent")
+  val GEOT_SIMPLE_CONTENT = new NameImpl("simpleContent")
+  val USERDATAITEM_SCHEMA = SchemaBuilder.record("userDataItem")
+    .namespace(AVRO_NAMESPACE)
+    .fields()
+    .name("class").`type`.stringType().noDefault()
+    .name("key").`type`.stringType().noDefault()
+    .name("value").`type`.stringType().noDefault().endRecord()
+
+  trait FieldNameEncoder {
+    def encode(name:String): String

```

Comment:
(style)  Unfortunately we haven't documented the GeoMesa code style documented (we do vaguely point at http://docs.scala-lang.org/style).  Here are a few things which would be nice to address:

1.  There's usually one space after colons and commas (i.e., ",", ":").
2.  There's usually a space on both sides of a production ("=>").
3.  For control structures, we almost never one-line them.  Despite the notes here (http://docs.scala-lang.org/style/control-structures), we tend to use {} for each branch.  It is verbose, but you don't get silly situations like this:

```
if (control);
foo
```
What does the above do:)?